agent/runner/actions/mongodb_explain_action.go

```diff
@@ -89,28 +98,156 @@ func (a *mongodbExplainAction) Run(ctx context.Context) ([]byte, error) {
 	}
 	defer client.Disconnect(ctx) //nolint:errcheck
 
-	var eq proto.ExampleQuery
+	return explainForQuery(ctx, client, a.params.Query)
+}
+
+func (a *mongodbExplainAction) sealed() {}
+
+func (e explain) prepareCommand() (bson.D, error) {
+	command := e.Command
+
+	switch e.Op {
+	case "query":
+		if len(command) == 0 {
+			command = e.Query
+		}
+
+		if len(command) == 0 || command[0].Key != "find" {
+			var filter any
+			if len(command) != 0 && command[0].Key == "query" {
+				filter = command[0].Value
+			} else {
+				filter = command
+			}
+
+			return bson.D{
+				{Key: "find", Value: e.getCollection()},
+				{Key: "filter", Value: filter},
+			}, nil
+		}
```

Comment:
I believe, this whole block is difficult to reason about when it comes to boolean logic, so it could be problematic to test it reliably.

Can we refactor it to make the logic a bit more straightforward?

Reply:
I already refactored this one from Percona Tools.

Original one:
`	case "query":
		if len(cmd) == 0 {
			cmd = self.Query
		}

		// MongoDB 2.6:
		//
		// "query" : {
		//   "query" : {
		//
		//   },
		//	 "$explain" : true
		// },
		if _, ok := cmd.Map()["$explain"]; ok {
			cmd = bson.D{
				{"explain", ""},
			}
			break
		}

		if len(cmd) == 0 || cmd[0].Key != "find" {
			var filter interface{}
			if len(cmd) > 0 && cmd[0].Key == "query" {
				filter = cmd[0].Value
			} else {
				filter = cmd
			}

			coll := ""
			s := strings.SplitN(self.Ns, ".", 2)
			if len(s) == 2 {
				coll = s[1]
			}

			cmd = bson.D{
				{"find", coll},
				{"filter", filter},
			}
		} else {
			for i := range cmd {
				switch cmd[i].Key {
				// PMM-1905: Drop "ntoreturn" if it's negative.
				case "ntoreturn":
					// If it's non-negative, then we are fine, continue to next param.
					if cmd[i].Value.(int64) >= 0 {
						continue
					}
					fallthrough
				// Drop $db as it is not supported in MongoDB 3.0.
				case "$db":
					if len(cmd)-1 == i {
						cmd = cmd[:i]
					} else {
						cmd = append(cmd[:i], cmd[i+1:]...)
					}
				}
			}
		}`
		
		
I will try think about what else could be simplified/refactored.